File: Homework-1/an_nguyen_task3.py
import sys
import json
import time
from typing import List
from pyspark.sql import SparkSession


def average(a_list: List[int]) -> float:
    return sum(a_list) / len(a_list)


def main(review_input, business_input, output_one, output_two):
    spark = SparkSession.builder.master("local[*]").appName("HW1").getOrCreate()
    reviews = spark             \
        .sparkContext           \
        .textFile(review_input) \
        .coalesce(8)            \
        .map(json.loads)        \
        .map(lambda x: (x["business_id"], x["stars"]))
    businesses = spark              \
        .sparkContext               \
        .textFile(business_input)   \
        .coalesce(8)                \
        .map(json.loads)            \
        .map(lambda x: (x["business_id"], x["state"]))
    joined_data = reviews                               \
        .join(businesses)                               \
        .coalesce(2)                                    \
        .map(lambda x: [x[1][1], x[1][0]])              \
        .groupByKey()                                   \
        .map(lambda x: [x[0], average(x[1])])           \
        .sortBy(lambda x: (-x[1], x[0]))

    """ Task B """
    result = {}
    #   Method 1: Collect all the data, and then print the first 5 states
    start = time.time()
    method_one_data = joined_data.collect()
    for d in method_one_data[:5]:
        print(d)
    process_time = time.time() - start
    result["m1"] = process_time
    #   Method 2: Take the first 5 states, and then print them
    start = time.time()
    method_two_data = joined_data.take(5)
    for d in method_two_data:
        print(d)
    process_time = time.time() - start
    result["m2"] = process_time
    #   Explanation:
    result["explanation"] = "The first() method will be faster since it only return a portion of the RDD (in this" + \
                            " case is 5) comparing to the collect() which will return everything. Hence, Method B" + \
                            " will take a lot less time than Method A."
    with open(output_two, "w", encoding="utf-8") as f:
        json.dump(result, f)

    """ Task A """
    with open(output_one, "w", encoding="utf-8") as f:
        f.write("state,stars\n")
        for item in method_one_data:
            f.write(f"{item[0]},{item[1]}\n")


if __name__ == "__main__":
    main(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4])

File: Homework-1/an_nguyen_task3.scala
import java.io.{File, PrintWriter}
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.SparkSession


object an_nguyen_task3 {
	implicit val formats: DefaultFormats.type = DefaultFormats

	case class Business(business_id: String, name: String, address: String, city: String, state: String,
						postal_code: String, latitude: Float, longitude: Float, stars: Float, review_count: Int,
						is_open: Int, attributes: Any, categories: Any, hours: Any)

	case class Review(review_id: String, user_id: String, business_id: String, stars: Int, date: String, text: String,
					  useful: Int, funny: Int, cool: Int)

	def main(args: Array[String]): Unit = {
		Logger.getLogger("org").setLevel(Level.INFO)
		val spark = SparkSession
			.builder()
			.appName("an_nguyen_task1")
			.config("spark.master", "local[*]")
			.config("spark.executor.memory", "6g")
			.getOrCreate()
		val reviews = spark
			.sparkContext
			.textFile(args(0))
			.coalesce(8)
			.map(convertToReview)
			.map(x => (x.business_id, x.stars))
		val business = spark
			.sparkContext
			.textFile(args(1))
			.coalesce(8)
			.map(convertToBusinesss)
			.map(x => (x.business_id, x.state))
		val joinedData = reviews
			.join(business)
			.coalesce(8)
			.map(x => (x._2._2, x._2._1))
			.groupByKey()
			.map(x => (x._1, x._2.sum.toFloat / x._2.size.toFloat))
			.sortBy(x => (-x._2, x._1))

		// Task B
		// 	Method 1
		var pw = new PrintWriter(new File(args(3)))
		pw.write("{")
		pw.write("\"m1\":")
		var start = System.currentTimeMillis()
		val method_one_data = joinedData.collect()
		for (i <- method_one_data.take(5)) {
			println(i)
		}
		var processTime = (System.currentTimeMillis() - start).toFloat / 1000.toFloat
		pw.write(processTime + ",")
		// 	Method 2
		pw.write("\"m2\":")
		start = System.currentTimeMillis()
		val method_two_data = joinedData.take(5)
		for (i <- method_two_data) {
			println(i)
		}
		processTime = (System.currentTimeMillis() - start).toFloat / 1000.toFloat
		pw.write(processTime + ",")
		//	Explanation
		pw.write("\"explanation\": \"The first() method will be faster since it only return a portion of the RDD " +
			"(in this case is 5) comparing to the collect() which will return everything. Hence, Method B will " +
			"take a lot less time than Method A.\"}")
		pw.close()

		// Task A
		pw = new PrintWriter(new File(args(2)))
		pw.write("state,starts\n")
		for (i <- method_one_data) {
			pw.write(i._1 + "," + i._2 + "\n")
		}
		pw.close()
	}

	def convertToReview(line: Any): Review = {
		val temp = line + ""
		parse(temp).extract[Review]
	}

	def convertToBusinesss(line: Any): Business = {
		val temp = line + ""
		parse(temp).extract[Business]
	}
}
